### ISSUE TITLE
The "fresh methods" restriction does not achieve its goal.

### ISSUE THREAD

**Issue by apblack** [gracelang/language#21]
Grace currently has the restriction that the object being inherited must be "fresh", that is, it is returned as the direct result of a method-request.  That request is usually on a class's method or on a factory method, but can in fact be a request on any method that tail-returns an object.

The students who wrote the _graphix_  library structured it as follows.  

```grace
method addCircle {
    // Create and return new circle shape, after adding it to self
    def circle = object {
         inherits shape
         ...  defines methods of a circle ...
    }
    shapes.add(circle)
    circle
}
```

and similarly for all of the other shapes.  This worked fine, but the resulting circles could not be inherited, because the addCircle method did not return a fresh object, as defined above.

This was a problem — for example, I wanted `apple` to inherit from `addCircle`.  So I got the students to change the structure of the library, so that it now reads as follows.

``` grace
factory method addCircle {
    // Create and return new circle shape, after adding it to self
    inherits shape
     ...  defines methods of a circle ...
    shapes.add(self)
}
```

This achieves _exactly the same thing_ as the original code, but meets the fresh-object restriction.

The fresh object restriction was designed to prevent a reference to the inherited object from "escaping" before the inheritance happens.  Such an externally-visible reference can be seen to change its type when the super-object is extended using inheritance.

The problem, as illustrated  by this example, is that the fresh-object restriction does not achieve its goal.  Because the initialization code can refer to `self`, references to the super-object can escape.

I don't _want_ to see objects changing their type.  I don't _want_ to see immutable objects mutating.  
Is there some other syntactic restriction that will achieve the goal?  Or should we admit that the "mutating object" semantics for inheritance was a _bad idea_, consign it to James's _bad idea_ list, and adopt a delegation-based semantics for inheritance, where the sub-object retains its own identity?

**Comment by KimBruce** [comment]
I believe one of the main issues for "fresh objects" was for running
initialization code at the right time.  As you point out, it does not stop
references to self from escaping the context before fully initialized.  I
believe we were well aware of that -- at least I was -- and was willing to
make that trade-off.  It was the programmer's responsibility to makes sure
it wasn't exposed at the wrong time.  In the objectdraw library, the last
step in constructing a ("final") graphic object was to add itself to the
canvas.

My own strong feeling is that we need something like traditional class
inheritance if we are going to get buy-in from instructors.  I'm less
concerned with what we do with "inheritance" on objects.  (I'm trying to
remember if I've ever inherited from an object -- though I know I have
defined objects by inheriting from classes.)  Someone pointed out in an
earlier discussion that in languages without inheritance, users invent it.
It may be possible to have object delegation as the primitive and class
inheritance as a derived notion, and might be worth while investigating.
However, for me the most important notion is support for class inheritance
with a reasonable notion of initialization.

**Comment by kjx** [comment]
> I believe one of the main issues for "fresh objects" was for running
> initialization code at the right time.

Actually I think it is mostly managing self. Michael's proposal for inherits statements anywhere is about managing initialisation. The catch is that there is no "right time" in general of objects can do up calls and down calls during inheritance. 

> As you point out, it does not stop
> references to self from escaping the context before fully initialized.

Yep. It just ensured the "right" self escapes.  

> My own strong feeling is that we need something like traditional class
> inheritance if we are going to get buy-in from instructors.

If we expect to replace Java or Python then yes.  If JavaScript, not so much.

> Someone pointed out in an
> earlier discussion that in languages without inheritance, users invent it.

Me:  Lua, JS, Tcl, Self...

> It may be possible to have object delegation as the primitive and class
> inheritance as a derived notion, and might be worth while investigating.

I think we've been investigating it for years - and it turns its the answer is that it's not- particularly because of the binding of Self during initialisation.  If super objects exist or are shared, then self can out be unified during initialisation.

Quite why it took 20 years to find this out I don't know. JS lives without this, I think...

> However, for me the most important notion is support for class inheritance
> with a reasonable notion of initialization

And there's the rub. 

J

**Comment by apblack** [comment]
So for me, a simple language with an absence of special cases is more important than mimicking Java.
Although this may look like an expedient decision today, how will it look in 10 years time?   If *all* existing languages did the same thing, the situation might be clearer.   But there is no clear "standard" to follow.

**Comment by KimBruce** [comment]
An important point that we have been making is that learning Grace should
make it easy to transition to other languages in wide usage.  We have to
convince faculty and students of this if we are going to get buy-in.  I
already see lots of hesitation around here (and elsewhere) to stepping out
to a language that is not mainstream.

I'm not saying this requires exactly Java semantics (or Python) for that
matter, but it should be an (obviously) easy transition or we won't get any
buy-in.

**Comment by kjx** [comment]
> So for me, a simple language with an absence of special cases is more important than mimicking Java.

absolutely


> adopt a delegation-based semantics for inheritance, where the sub-object retains its own identity?

if not now, then when?

**Comment by apblack** [comment]
> if not now, then when?
Probably only once Kim has retired :-)   Until then, I'll probably teach traits but not inheritance.

**Comment by kjx** [comment]
Ahh OK.  especially as Kim doesn't seem to read these comments

**Comment by kjx** [comment]
FWIW I think I'd teach both, I remember typing some comments last night about how to use each.

HMM: another option for Europlop?

**Comment by KimBruce** [comment]
I do read them, but try only to reply when I have something useful to say.

If you leave out inheritance you will drop 80% of those who might otherwise
use Grace. On the other hand 20% of a very small number is nearly as small.

**Comment by kjx** [comment]
> On 13/02/2016, at 15:40pm, Kim Bruce <notifications@github.com> wrote:
> 
> I do read them, but try only to reply when I have something useful to say.

well OK that's good!

> If you leave out inheritance you will drop 80% of those who might otherwise
> use Grace. On the other hand 20% of a very small number is nearly as small.

:-)

I like a model where traits are used “in the large”, typically in broad and shallow hierarchies,
and where inheritance is used for extending frameworks and conceptual modelling.

J

**Comment by apblack** [comment]
I don't think anyone was proposing that we take inheritance out of Grace.

I'm going to close this thread because the claim that caused me to open it is false, and because this is not a productive discussion.
